Allocate megapages from different page provider than large allocations in libpas

ddegazio opened the pull request:
#### 968212ccdd07944a158b38d628827e06f2f1fa89
<pre>
Allocate megapages from different page provider than large allocations in libpas
<a href="https://bugs.webkit.org/show_bug.cgi?id=284338">https://bugs.webkit.org/show_bug.cgi?id=284338</a>
<a href="https://rdar.apple.com/138070544">rdar://138070544</a>

Reviewed by Yusuke Suzuki.

This patch separates out the way megapages are allocated from other
large allocations in libpas. First, we create a distinct bootstrap heap
and page provider for small/medium allocations. Next, we create a new
large free heap whose page provider is this new bootstrap heap. Next,
for each pas_heap, we create a new megapage_large_heap alongside the
existing large heap, a pas_large_heap backed by the new bootstrap heap.
Finally, we change the way megapage caches are constructed, to use this
new large heap as the allocator for megapages instead of self-hosting
using the existing large heap. Altogether, this means megapages, and the
small/medium objects within them, are no longer allocated out of the same
page source as marge/large objects. This patch also adds some new fields
to page configs and local allocators to make it clear whether a page or
allocator belongs to a megapage or not.

This patch also decreases the memory pressure in two OOM layout tests.
These tests are relying too strongly on the absence of an OOM error in
subsequent code, and with this patch perturbing the heap, it seems they
consistently but spuriously crash with an OOM error. Removing some
allocations in these tests reduces the chance we get an OOM after the
critical section of the test is over, preventing spurious failures.

* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/jit_heap_config.h:
* Source/bmalloc/libpas/src/libpas/pas_basic_heap_page_caches.h:
* Source/bmalloc/libpas/src/libpas/pas_bootstrap_free_heap.c:
(bootstrap_source_allocate_aligned):
* Source/bmalloc/libpas/src/libpas/pas_compact_heap_reservation.c:
(pas_compact_heap_reservation_try_allocate):
* Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
(allocate_from_megapages):
(pas_create_basic_heap_page_caches_with_reserved_memory):
* Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.c:
(pas_enumerable_page_malloc_try_allocate_without_deallocating_padding):
* Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h:
* Source/bmalloc/libpas/src/libpas/pas_enumerator_region.c:
(pas_enumerator_region_allocate):
* Source/bmalloc/libpas/src/libpas/pas_heap.c:
(pas_heap_create):
* Source/bmalloc/libpas/src/libpas/pas_heap.h:
(pas_heap_for_large_heap):
* Source/bmalloc/libpas/src/libpas/pas_heap_config_utils.c:
(pas_heap_config_utils_allocate_aligned):
* Source/bmalloc/libpas/src/libpas/pas_heap_config_utils.h:
* Source/bmalloc/libpas/src/libpas/pas_heap_kind.h:
(pas_heap_kind_get_string):
* Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
(pas_large_heap_construct):
* Source/bmalloc/libpas/src/libpas/pas_large_heap.h:
* Source/bmalloc/libpas/src/libpas/pas_large_heap_physical_page_sharing_cache.h:
* Source/bmalloc/libpas/src/libpas/pas_local_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_set_up_bump):
(pas_local_allocator_make_bump):
(pas_local_allocator_set_up_primordial_bump):
(pas_local_allocator_try_allocate_inline_cases):
* Source/bmalloc/libpas/src/libpas/pas_megapage_cache.h:
* Source/bmalloc/libpas/src/libpas/pas_page_base_config.h:
* Source/bmalloc/libpas/src/libpas/pas_page_config_size.h: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h.
(pas_page_config_size_get_string):
* Source/bmalloc/libpas/src/libpas/pas_page_malloc.c:
(pas_page_malloc_try_map_pages):
(pas_page_malloc_try_allocate_without_deallocating_padding):
* Source/bmalloc/libpas/src/libpas/pas_page_malloc.h:
* Source/bmalloc/libpas/src/libpas/pas_segregated_page_config_kind.h:
(pas_segregated_page_config_kind_is_small):
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_free_heap.c: Copied from Source/bmalloc/libpas/src/libpas/pas_bootstrap_free_heap.c.
(bootstrap_source_allocate_aligned):
(initialize_config):
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_free_heap.h: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h.
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_heap_page_provider.c: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.c.
(pas_small_medium_bootstrap_heap_page_provider):
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_heap_page_provider.h: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h.
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_intrinsic.h:
* Source/bmalloc/libpas/src/libpas/pas_utility_heap_config.h:

Canonical link: <a href="https://commits.webkit.org/288057@main">https://commits.webkit.org/288057@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d2f3a581bb99dd1b399523849f3896cb4b7a5ece

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81752 "47 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35706 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86297 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32748 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1311 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9098 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63777 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21500 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84822 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/952 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74404 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44063 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/852 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28579 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31201 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74732 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29182 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87735 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80806 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6388 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72113 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9177 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70224 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71343 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17776 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15441 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14358 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103218 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8943 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25062 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8784 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12307 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10592 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator.h:
```diff
@@ -55,6 +55,7 @@ struct pas_local_allocator {
     uint8_t alignment_shift;
     pas_local_allocator_config_kind config_kind : 8;
     bool current_word_is_valid; /* This is just used by enumeration. */
+    bool is_small_bumpable; /* Marks that the bumpable region in this local allocator is part of a small page. */
```
Is it possible to determine this from the object_size somehow, rather than having a second source of truth?

Achierius replied:
Alternatively, having it link to the page_config (since we always set this bool according to `page_config.base.page_config_size == pas_page_config_size_small`) -- that's an extra 7B but this is already a pretty large structure so I'd guess that it wouldn't matter

ddegazio replied:
I'm not 100% sure why, to be honest, but it doesn't appear that object_size is sufficient to identify small v.s. medium. The page config _is_ sufficient, but in the bump-allocating case, we don't have it available - we can materialize it, but only by `switch`ing on the local allocator kind. So this bool lets us remember that detail of the config kind from when we initially construct the local allocator, then query it more easily in the bump allocating path.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_page_malloc.c:
```diff
@@ -86,11 +86,11 @@ PAS_NEVER_INLINE size_t pas_page_malloc_alignment_shift_slow(void)
 
 static void*
-pas_page_malloc_try_map_pages(size_t size)
+pas_page_malloc_try_map_pages(size_t size, bool may_contain_small_or_medium)
```
What does this bool actually denote -- whether the page-malloc might eventually hold objects in the small/medium size classes? Or more specifically whether it will be used as backing storage for a small/medium heap?

If it's the latter, perhaps it'd be cleaner to have a second function which specifically allocates backing page memory for the backing memory of heaps, and then have that take in a size-class parameter. As-is you have to pass in `false` in a lot of places that don't really care about size-classes at all (as opposed to the ones which are doing so because they are using it as backing storage for a marge/large heap)

ddegazio replied:
Resolved offline - it's the latter, but at the moment there's not much compelling reason to separate the two out.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_page_config_size.h:
```diff
@@ -0,0 +1,61 @@
+
+PAS_BEGIN_EXTERN_C;
+
+enum pas_page_config_size {
```
nit: since this isn't really the size of the page-config per se, but the objects allocated underneath it, could it have a different name? Perhaps even 'pas_object_size_class' since it's supposed to be 1-to-1 with that concept (currently just macros)

ddegazio replied:
I think maybe `pas_page_config_size_category` is good? I like keeping it associated with the `page_config` because it's not really any individual object's size either - small/medium/marge/large is decided at the page level. But you're right that it's not really the size per se.

ddegazio replied:
Resolved in latest push.

Constellation reviewed (approved):
rs=me

ddegazio committed:
Allocate megapages from different page provider than large allocations in libpas
https://bugs.webkit.org/show_bug.cgi?id=284338
rdar://138070544

Reviewed by Yusuke Suzuki.

This patch separates out the way megapages are allocated from other
large allocations in libpas. First, we create a distinct bootstrap heap
and page provider for small/medium allocations. Next, we create a new
large free heap whose page provider is this new bootstrap heap. Next,
for each pas_heap, we create a new megapage_large_heap alongside the
existing large heap, a pas_large_heap backed by the new bootstrap heap.
Finally, we change the way megapage caches are constructed, to use this
new large heap as the allocator for megapages instead of self-hosting
using the existing large heap. Altogether, this means megapages, and the
small/medium objects within them, are no longer allocated out of the same
page source as marge/large objects. This patch also adds some new fields
to page configs and local allocators to make it clear whether a page or
allocator belongs to a megapage or not.

This patch also decreases the memory pressure in two OOM layout tests.
These tests are relying too strongly on the absence of an OOM error in
subsequent code, and with this patch perturbing the heap, it seems they
consistently but spuriously crash with an OOM error. Removing some
allocations in these tests reduces the chance we get an OOM after the
critical section of the test is over, preventing spurious failures.

* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/jit_heap_config.h:
* Source/bmalloc/libpas/src/libpas/pas_basic_heap_page_caches.h:
* Source/bmalloc/libpas/src/libpas/pas_bootstrap_free_heap.c:
(bootstrap_source_allocate_aligned):
* Source/bmalloc/libpas/src/libpas/pas_compact_heap_reservation.c:
(pas_compact_heap_reservation_try_allocate):
* Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
(allocate_from_megapages):
(pas_create_basic_heap_page_caches_with_reserved_memory):
* Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.c:
(pas_enumerable_page_malloc_try_allocate_without_deallocating_padding):
* Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h:
* Source/bmalloc/libpas/src/libpas/pas_enumerator_region.c:
(pas_enumerator_region_allocate):
* Source/bmalloc/libpas/src/libpas/pas_heap.c:
(pas_heap_create):
* Source/bmalloc/libpas/src/libpas/pas_heap.h:
(pas_heap_for_large_heap):
* Source/bmalloc/libpas/src/libpas/pas_heap_config_utils.c:
(pas_heap_config_utils_allocate_aligned):
* Source/bmalloc/libpas/src/libpas/pas_heap_config_utils.h:
* Source/bmalloc/libpas/src/libpas/pas_heap_kind.h:
(pas_heap_kind_get_string):
* Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
(pas_large_heap_construct):
* Source/bmalloc/libpas/src/libpas/pas_large_heap.h:
* Source/bmalloc/libpas/src/libpas/pas_large_heap_physical_page_sharing_cache.h:
* Source/bmalloc/libpas/src/libpas/pas_local_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_set_up_bump):
(pas_local_allocator_make_bump):
(pas_local_allocator_set_up_primordial_bump):
(pas_local_allocator_try_allocate_inline_cases):
* Source/bmalloc/libpas/src/libpas/pas_megapage_cache.h:
* Source/bmalloc/libpas/src/libpas/pas_page_base_config.h:
* Source/bmalloc/libpas/src/libpas/pas_page_config_size.h: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h.
(pas_page_config_size_get_string):
* Source/bmalloc/libpas/src/libpas/pas_page_malloc.c:
(pas_page_malloc_try_map_pages):
(pas_page_malloc_try_allocate_without_deallocating_padding):
* Source/bmalloc/libpas/src/libpas/pas_page_malloc.h:
* Source/bmalloc/libpas/src/libpas/pas_segregated_page_config_kind.h:
(pas_segregated_page_config_kind_is_small):
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_free_heap.c: Copied from Source/bmalloc/libpas/src/libpas/pas_bootstrap_free_heap.c.
(bootstrap_source_allocate_aligned):
(initialize_config):
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_free_heap.h: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h.
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_heap_page_provider.c: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.c.
(pas_small_medium_bootstrap_heap_page_provider):
* Source/bmalloc/libpas/src/libpas/pas_small_medium_bootstrap_heap_page_provider.h: Copied from Source/bmalloc/libpas/src/libpas/pas_enumerable_page_malloc.h.
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_intrinsic.h:
* Source/bmalloc/libpas/src/libpas/pas_utility_heap_config.h:

Canonical link: https://commits.webkit.org/288057@main
